client/api/src/proof_provider.rs

```diff
@@ -0,0 +1,67 @@
+// Copyright 2017-2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+//! Proof utilities
+use sp_runtime::{
+	generic::BlockId,
+	traits::{Block as BlockT},
+};
+use crate::{StorageProof, ChangesProof};
+use sp_storage::{ChildInfo, StorageKey};
+
+/// Interface for providing block proving utilities.
+pub trait ProofProvider<Block: BlockT> {
+	/// Reads storage value at a given block + key, returning read proof.
+	fn read_proof(&self, id: &BlockId<Block>, keys: &[Vec<u8>]) -> sp_blockchain::Result<StorageProof>;
```

Comment:
It's unfortunate that we'll have to do temporary allocations for `keys` here. Maybe we could declare it like this?
```
fn read_proof(&self, id: &BlockId<Block>, keys: &mut dyn Iterator<Item=&[u8]>)
```
and then use `map` at call site to get `&[u8]` where needed?

Reply:
So I took a look at it and unfortunately, the `read_proof` is being used in contexts where a mutable reference to `keys` isn't available.

Reply:
This is a reference to iterator, not to keys which has to be mutable because `next` method for an iterator takes `&mut self`

Reply:
```
let keys = [[1,2], [3, 4]];
x.read_proof(id, &mut key.iter().map(AsRef::as_ref)))
```
This should work even though the data itself is not mutable

Reply:
> This is a reference to iterator, not to keys

this is an example.

https://github.com/paritytech/substrate/blob/b5ec7d412d27feafad79a7c94b6f6451f0efc04f/client/network/src/protocol/light_client_handler.rs#L475